[Docs][k8s] Revamp k8s cluster admin docs (#3649)
skypilot-org/skypilot: docs/source/reference/kubernetes/kubernetes-setup.rst

    explains how to submit tasks to your cluster.

To prepare a Kubernetes cluster to run SkyPilot, the cluster administrator must:

1. :ref:`Deploy a cluster <kubernetes-setup-deploy>` running Kubernetes v1.20 or later.
2. Set up :ref:`GPU support <kubernetes-setup-gpusupport>`.
3. [Optional] :ref:`Set up ports <kubernetes-setup-ports>` for exposing services.
4. [Optional] :ref:`Set up permissions <kubernetes-setup-serviceaccount>`: Create a namespace for your users and/or create a service account with minimal permissions for SkyPilot.

> After these steps, the administrator can share the kubeconfig file with users, who can then submit tasks to the cluster using SkyPilot.

Reviewer: I would prefer to mention the last three sections `verifying setup`, `observability`, and `troubleshooting` at the beginning as well. Otherwise, they are quite buried below.

Author: Good idea - I've added a toc at the top. Not including troubleshooting since it is mentioned on the index page and appears on both left and right sidebars. wdyt?

    .. grid-item-card::  On-prem Clusters (RKE2, K3s, etc.)
        :link: kubernetes-setup-onprem
        :link-type: ref
        :text-align: center

        For on-prem deployments with kubeadm, RKE2, K3s or other distributions.

    .. grid-item-card:: Google Cloud - GKE
        :link: kubernetes-setup-gke
        :link-type: ref
        :text-align: center

        Google's hosted Kubernetes service.

    .. grid-item-card::  Amazon - EKS
        :link: kubernetes-setup-eks
        :link-type: ref
        :text-align: center

        Amazon's hosted Kubernetes service.

.. _kubernetes-setup-gpusupport:

Step 2 - Set up GPU support
---------------------------

To utilize GPUs on Kubernetes, your cluster must:

1. Have the ``nvidia.com/gpu`` **resource** available on all GPU nodes and have ``nvidia`` as the default runtime for your container engine. See `Nvidia GPU Operator Installation Guide <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#install-nvidia-gpu-operator>`_ for more details.
2. Have a **label on each node specifying the GPU type**. See :ref:`Setting up GPU labels <kubernetes-gpu-labels>` for more details.

> .. tip::

Reviewer: Can we move this tip right after number 1?

Author: I'm leaning towards keeping the tips/notes after the list of requirements, otherwise it gets hard to read what all needs to be done to setup GPU support.

Reviewer: That makes sense! Then can we add a sentence in number 1 to say that clusters setup with the deployment instructions do not need to setup the gpu operator? So a user does not need to go through the nvidia doc only to find out it is not needed for their cluster?

Also, we may want to mention that the instructions of gpu operators setup are provided in the deployment doc, as the nvidia doc might be too long to follow.

Author: Good point - added a sub bullet. does that help?

>     You can check if Nvidia GPU operator is installed and the ``nvidia`` runtime is set as default by running:

Reviewer: ```suggestion
    If your kubernetes cluster is set up with the instructions in Step 1, the GPU operator should have been set up. You can check if Nvidia GPU operator is installed and the ``nvidia`` runtime is set as default by running:
```

Author: Reworded a little - is it better now?

To use GPUs with SkyPilot, cluster nodes must be labelled with the GPU type. This informs SkyPilot which GPU types are available on the cluster.

Currently supported labels are:

* ``nvidia.com/gpu.product``: automatically created by Nvidia GPU Operator.
* ``cloud.google.com/gke-accelerator``: used by GKE clusters.
* ``karpenter.k8s.aws/instance-gpu-name``: used by Karpenter.
* ``skypilot.co/accelerator``: custom label used by SkyPilot if none of the above are present.

Any one of these labels is sufficient for SkyPilot to detect GPUs on the cluster.

.. tip::

    To check if your nodes contain the necessary labels, run:

    .. code-block:: bash

        output=$(kubectl get nodes --show-labels | awk -F'[, ]' '{for (i=1; i<=NF; i++) if ($i ~ /nvidia.com\/gpu.product=|cloud.google.com\/gke-accelerator=|karpenter.k8s.aws\/instance-gpu-name=|skypilot.co\/accelerator=/) print $i}')
        if [ -z "$output" ]; then
          echo "No valid GPU labels found."
        else
          echo "GPU Labels found:"
          echo "$output"
        fi

> Manually Labelling Nodes

Reviewer: Let's have `Automatically Labelling Nodes` before this `Manually Labelling Nodes`, so users do not have to manually labelled all the nodes and then run into the next section for automatic labelling?

    If the GPU labelling process fails, you can run ``python -m sky.utils.kubernetes.gpu_labeler --cleanup`` to clean up the failed jobs.

.. _kubernetes-setup-ports:

[Optional] Step 3 - Set up for Exposing Services
------------------------------------------------

.. tip::

>     If you are using GKE or EKS or do not plan to run SkyServe on Kubernetes, no additional setup is required to expose ports. SkyPilot will create a LoadBalancer service automatically.

Reviewer: ```suggestion
    If you are using GKE or EKS or do not plan to expose ports publicly on Kubernetes (such as ``sky launch --ports``, SkyServe), no additional setup is required to expose ports. SkyPilot will create a LoadBalancer service automatically.
```

    `Kubernetes documentation <https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/>`_
    for more information on how to set up access control for the dashboard.

Troubleshooting Kubernetes Setup
--------------------------------

If you encounter issues while setting up your Kubernetes cluster, please refer to the :ref:`troubleshooting guide <kubernetes-troubleshooting>` to diagnose and fix issues.

.. toctree::
   :hidden:

   kubernetes-deployment

>    kubernetes-ports

Reviewer: How about reduce the length a bit, so that it does change lines in the TOC?
```suggestion
   Exposing Services <kubernetes-ports>
```

Also, would be nice to change the title in TOC for `Getting Started on Kubernetes` to `Getting Started`?
